src/unistd.rs

```diff
@@ -2136,3 +2136,201 @@ mod setres {
+                dir: CString::new(CStr::from_ptr(pw.pw_dir).to_bytes()).unwrap(),
+                shell: CString::new(CStr::from_ptr(pw.pw_shell).to_bytes()).unwrap(),
+                uid: Uid::from_raw(pw.pw_uid),
+                gid: Gid::from_raw(pw.pw_gid),
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                class: CString::new(CStr::from_ptr(pw.pw_class).to_bytes()).unwrap(),
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                change: pw.pw_change,
+                #[cfg(not(any(target_os = "linux", target_os = "android")))]
+                expire: pw.pw_expire
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Group {
+    /// group name
+    pub name: CString,
+    /// group ID
+    pub gid: Gid,
+    /// list of members
+    pub mem: Vec<CString>
+}
+
+impl From<libc::group> for Group {
+    fn from(gr: libc::group) -> Group {
+        unsafe {
+            Group {
+                name: CString::new(CStr::from_ptr(gr.gr_name).to_bytes()).unwrap(),
+                gid: Gid::from_raw(gr.gr_gid),
+                mem: Group::members(gr.gr_mem)
+            }
+        }
+    }
+}
+
+impl Group {
+    unsafe fn members(mem: *mut *mut c_char) -> Vec<CString> {
+        let mut ret = vec![];
+
+        for i in 0.. {
+            let u = mem.offset(i);
+            if (*u).is_null() {
+                break;
+            } else {
+                let cs = CString::new(CStr::from_ptr(*u).to_bytes()).unwrap();
+                ret.push(cs);
+            }
+        }
+
+        ret
+    }
+}
+
+#[inline]
+/// Gets a user DB entry based on username (see
+/// [getpwnam(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwnam.html))
+pub fn getpwnam<S: Into<Vec<u8>>>(name: S) -> Result<Option<Passwd>> {
+    let pw = unsafe { libc::getpwnam(CString::new(name).unwrap().as_ptr()).as_ref() };
+
+    if errno::errno() != 0 {
+        return Err(Error::Sys(Errno::last()));
+    }
+
+    match pw {
+        None => {return Ok(None)},
+        Some(pw) => Ok(Some(Passwd::from(*pw)))
+    }
+}
+
+/// Gets a user DB entry based on user's ID (see
+/// [getpwuid(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid.html))
+#[inline]
+pub fn getpwuid(uid: Uid) -> Result<Option<Passwd>> {
+    let pw = unsafe { libc::getpwuid(uid.0).as_ref() };
+
+    if errno::errno() != 0 {
+        return Err(Error::Sys(Errno::last()));
+    }
+
+    match pw {
+        None => {return Ok(None)},
+        Some(pw) => Ok(Some(Passwd::from(*pw)))
+    }
+}
+
+/// Gets a user DB entry (see
+/// [getpwent(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwent.html))
+#[cfg(not(target_os = "android"))]
+#[inline]
+pub fn getpwent() -> Result<Option<Passwd>> {
+    let pw = unsafe { libc::getpwent().as_ref() };
+
+    if errno::errno() != 0 {
```

Comment:
Is it smart to check the `errno` in this way?

Reply:
It works, but I do have to admit that I did not really understand too well the `errno.rs` module.

Reply:
`libc::getpwent` also uses a static object.  It also isn't very Rusty, because it's basically the `next` method for an implied iterator.  I don't think it should be directly exposed by Nix.  Instead, I think you should create an iterator class that will internally use `libc::getpwent_r` to iterate through the password file.  Something like this:
```rust
struct PwEntries {
    ...
}
impl PwEntries {
    fn new() -> Self {
        libc::setpwent();
        ...
    }
}
impl Iterator for PwEntries {
    type item = Passwd;
    fn next(&mut self) -> Option<Passwd> {
        ...
    }
}
impl Drop for PwEntries {
    fn drop(&mut self) {
        libc::endpwent();
    }
}
```
Unfortunately, this still isn't 100% reentrant since those `setpwent` and `endpwent` functions still store their file descriptors in static variables.  But I don't see a better option.

Reply:
`libc::getpwent_r` does not exist, so I suppose this is on hold until I can figure out how to get it added to libc...